api/filters/replacement/replacement.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2019 The Kubernetes Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package replacement
+
+import (
+	"fmt"
+	"strings"
+
+	"sigs.k8s.io/kustomize/api/provider"
+	"sigs.k8s.io/kustomize/api/resmap"
+	"sigs.k8s.io/kustomize/api/types"
+	"sigs.k8s.io/kustomize/kyaml/kio"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+type Filter struct {
+	Replacements []types.Replacement
+}
+
+var _ kio.Filter = Filter{}
+
+// Filter replaces values of targets with values from sources
+// TODO (#3492): Connect this to a replacement transformer plugin
+func (f Filter) Filter(nodes []*yaml.RNode) ([]*yaml.RNode, error) {
+	var result []*yaml.RNode
+	rf := provider.NewDefaultDepProvider().GetResourceFactory()
+	m, err := resmap.NewFactory(rf).NewResMapFromRNodeSlice(nodes)
+	if err != nil {
+		return nil, err
+	}
+	for i := range nodes {
```

Comment:
I'm wondering about the order of the loops here... IIUC, it is possible for one replacement to change the value at the source of another. If that's true, then it matters whether we apply all the transformations to one node and then move onto the next node, vs. applying the first transformation to all nodes and then moving onto the next transformation. I think intuitively I'd expect the latter, whereas we're doing the former. We should probably have a test either way. WDYT?

Reply:
yeah that makes sense to me. I've rewritten it to apply one replacement at a time to all the nodes before moving onto the next one